drivers/platform/xilinx/timer.c

```diff
@@ -0,0 +1,385 @@
+
+	xil_timer_dev->timer_counter_intr = 0;
+	switch(xil_timer_dev->type) {
+	case TIMER_PL:
+#ifdef XTMRCTR_H
+		XTmrCtr_Start(xil_timer_dev->instance, xil_timer_dev->active_tmr);
+
+		break;
+#endif
+	case TIMER_PS:
+#ifdef XSCUTIMER_H
+		XScuTimer_Start(xil_timer_dev->instance);
+
+		break;
+#endif
+	default:
+		return FAILURE;
+	}
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Stop a timer from counting.
+ * @param [in] desc - Pointer to the device handler.
+ * @return 0 in case of success, or negative error code in the following cases:
+ *         - the timer type is wrong.
+ */
+int32_t timer_stop(struct timer_desc *desc)
+{
+	struct xil_timer_desc *xil_timer_dev = desc->extra;
+
+	switch(xil_timer_dev->type) {
+	case TIMER_PL:
+#ifdef XTMRCTR_H
+		XTmrCtr_Stop(xil_timer_dev->instance, xil_timer_dev->active_tmr);
+
+		break;
+#endif
+	case TIMER_PS:
+#ifdef XSCUTIMER_H
+		XScuTimer_Stop(xil_timer_dev->instance);
+
+		break;
+#endif
+	default:
+		return FAILURE;
+	}
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Get the value of the counter register for the timer.
+ * @param [in]  desc    - Pointer to the device handler.
+ * @param [out] counter - Pointer to the counter value.
+ * @return 0 in case of success, or negative error code in the following cases:
+ *         - the timer type is wrong.
+ */
+int32_t timer_counter_get(struct timer_desc *desc, uint32_t *counter)
+{
+	struct xil_timer_desc *xil_timer_dev = desc->extra;
+
+	switch(xil_timer_dev->type) {
+	case TIMER_PL:
+#ifdef XTMRCTR_H
+		*counter = XTmrCtr_GetValue((XTmrCtr *)xil_timer_dev->instance,
+					    xil_timer_dev->active_tmr);
+
+		break;
+#endif
+	case TIMER_PS:
+#ifdef XSCUTIMER_H
+		*counter = XScuTimer_GetCounterValue((XScuTimer *)xil_timer_dev->
+						     instance);
+
+		break;
+#endif
+	default:
+		return FAILURE;
+	}
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Set the timer counter register value.
+ * @param [in] desc    - Pointer to the device handler.
+ * @param [in] new_val - The new value of the counter register.
+ * @return 0 in case of success, error code otherwise.
+ */
+int32_t timer_counter_set(struct timer_desc *desc, uint32_t new_val)
+{
+	if(desc) {
+
+	}
+
+	if(new_val) {
+
+	}
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Get the timer clock frequency.
+ * @param [in]  desc    - Pointer to the device handler.
+ * @param [out] freq_hz - The new value of the counter register.
+ * @return 0 in case of success, or negative error code in the following cases:
+ *         - the timer type is wrong.
+ */
+int32_t timer_count_clk_get(struct timer_desc *desc, uint32_t *freq_hz)
+{
+	struct xil_timer_desc *xil_timer_dev = desc->extra;
+
+	switch(xil_timer_dev->type) {
+	case TIMER_PL:
+#ifdef XTMRCTR_H
+		*freq_hz = XPAR_AXI_TIMER_CLOCK_FREQ_HZ;
+
+		break;
+#endif
+	case TIMER_PS:
+#ifdef XSCUTIMER_H
+		*freq_hz = XPAR_CPU_CORTEXA9_CORE_CLOCK_FREQ_HZ / 2;
+
+		break;
+#endif
+	default:
+		return FAILURE;
+	}
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Set the timer clock frequency.
```

Comment:
You should add in the comment that the clock frequency can't be changed.

Reply:
Will do